ddcommon/src/entity_id/unix/cgroup_inode.rs

```diff
@@ -0,0 +1,143 @@
+// Copyright 2021-Present Datadog, Inc. https://www.datadoghq.com/
+// SPDX-License-Identifier: Apache-2.0
+
+//! This module provides functions to fetch cgroup node path and fetching it's inode
+use super::CgroupFileParsingError;
+use std::fs::File;
+use std::io::{BufRead, BufReader};
+use std::os::unix::fs::MetadataExt;
+use std::path::{Path, PathBuf};
+use std::{fs, io};
+
+// Those two variables are unused in tests
+#[cfg(not(test))]
+// From https://github.com/torvalds/linux/blob/5859a2b1991101d6b978f3feb5325dad39421f29/include/linux/proc_ns.h#L41-L49
+// Currently, host namespace inode number are hardcoded, which can be used to detect
+// if we're running in host namespace or not (does not work when running in DinD)
+const HOST_CGROUP_NAMESPACE_INODE: u64 = 0xEFFFFFFB;
+
+#[cfg(not(test))]
+const DEFAULT_CGROUP_NS_PATH: &str = "/proc/self/ns/cgroup";
+
+/// Returns the inode of file at `path`
+fn get_inode(path: &Path) -> io::Result<u64> {
+    let meta = fs::metadata(path)?;
+    Ok(meta.ino())
+}
+
+/// Returns the cgroup mount path associated with `base_controller` or the default one for
+/// cgroupV2
+fn get_cgroup_node_path(
+    base_controller: &str,
+    cgroup_path: &Path,
+    cgroup_mount_path: &Path,
+) -> Result<PathBuf, CgroupFileParsingError> {
+    let file = File::open(cgroup_path).map_err(|_| CgroupFileParsingError::CannotOpenFile)?;
+    let reader = BufReader::new(file);
+
+    let mut node_path: Option<PathBuf> = None;
+
+    for (index, line) in reader.lines().enumerate() {
+        let line_content = &line.map_err(|_| CgroupFileParsingError::InvalidFormat)?;
+        let cgroup_entry: Vec<&str> = line_content.split(':').collect();
+        if cgroup_entry.len() != 3 {
+            return Err(CgroupFileParsingError::InvalidFormat);
+        }
+        let controllers: Vec<&str> = cgroup_entry[1].split(',').collect();
+        // Only keep empty controller if it is the first line as cgroupV2 uses only one line
+        if controllers.contains(&base_controller) || (controllers.contains(&"") && index == 0) {
```

Comment:
It's adding the assumption that the cgroup line will always have index 0 which should hold in practice but the Agent and other implementations don't have this assumption. IMO it's better to harmonize just in case the behavior ever changes.